app/src/main/java/org/mozilla/fenix/library/history/HistorySearchDialogFragment.kt

```diff
@@ -254,4 +264,48 @@ class HistorySearchDialogFragment : AppCompatDialogFragment(), UserInteractionHa
             }
         }
     }
+
+    private val startForResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+        result: ActivityResult ->
+        if (result.resultCode == Activity.RESULT_OK) {
+            val intent = result.data
+            intent?.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)?.first()?.also {
+                toolbarView.view.edit.updateUrl(url = it, shouldHighlight = true)
+                interactor.onTextChanged(it)
+                toolbarView.view.edit.focus()
+            }
+        }
+    }
+
+    private fun addVoiceSearchButton() {
+        if (voiceSearchButtonAlreadyAdded || !isVisible) return
+
+        toolbarView.view.addEditAction(
+            BrowserToolbar.Button(
+                imageDrawable = AppCompatResources.getDrawable(requireContext(), R.drawable.ic_microphone)!!,
+                contentDescription = requireContext().getString(R.string.voice_search_content_description),
+                visible = { true },
+                listener = ::launchVoiceSearch
+            )
+        )
+
+        voiceSearchButtonAlreadyAdded = true
+    }
+
+    private fun launchVoiceSearch() {
+        // Note if a user disables speech while the app is on the search fragment
+        // the voice button will still be available and *will* cause a crash if tapped,
+        // since the `visible` call is only checked on create. In order to avoid extra complexity
+        // around such a small edge case, we make the button have no functionality in this case.
+        if (!isSpeechAvailable()) { return }
+
+        speechIntent.apply {
+            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
+            putExtra(RecognizerIntent.EXTRA_PROMPT, requireContext().getString(R.string.voice_search_explainer))
+        }
+
+        startForResult.launch(speechIntent)
+    }
+
+    private fun isSpeechAvailable(): Boolean = speechIntent.resolveActivity(requireContext().packageManager) != null
```

Comment:
I see we have a settings for `shouldShowVoiceSearch`, can you check if other voice search utilizes this settings https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt#109

Reply:
I removed it thinking History search is not related to search settings.  But I think you're right.  Since the final goal is to unify them, I'll check the settings before showing the voice search button.  Good catch, Thanks